src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -281,11 +284,68 @@ RecordPlacementAccessToCache(Oid relationId, ShardPlacementAccessType accessType
 }
 
 
+/*
+ * RecordParallelAccessForTaskList gets a task list and records
+ * the necessary parallel relation accesses for the task list.
+ *
+ * This function is used to enforce foreign keys from distributed
+ * tables to reference tables.
+ */
+void
+RecordParallelAccessForTaskList(List *taskList)
+{
+	Task *firstTask = NULL;
+
+	if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+	{
+		/* sequential mode prevents parallel access */
+		return;
+	}
+
+	if (list_length(taskList) < 2)
+	{
+		/* single shard task doesn't mean parallel access in our definition */
+		return;
+	}
+
+	firstTask = linitial(taskList);
+
+	if (firstTask->taskType == SQL_TASK)
+	{
+		RecordRelationParallelSelectAccessForTask(firstTask);
```

Comment:
In the previous version of code, where this is done in `MultiRealTimeExecute()`, we loop over all tasks and call this function for each of the tasks. Is calling it only for the first task ok?

I'm assuming all tasks are on the same set of relations and this should be ok. just wanted to make sure.

Reply:
Yes, that's fine. In the end, we're only tracking relation accesses and each task accesses the same relations.